src/plugins/gui/providers/kadoOtcProvider.ts

```diff
@@ -0,0 +1,454 @@
+  amount: number
+  blockchain: string
+  currency: string
+  reverse: boolean
+  asset: string
+}
+
+export const kadoOtcProvider: FiatProviderFactory = {
+  providerId,
+  storeId,
+  makeProvider: async (params: FiatProviderFactoryParams): Promise<FiatProvider> => {
+    const { apiKeys } = params
+    const { apiKey, apiUserEmail } = asApiKeys(apiKeys)
+
+    const authToken = btoa(`${apiUserEmail}/token:${apiKey}`) // base64 encode this
+
+    const instance: FiatProvider = {
+      providerId,
+      partnerIcon,
+      pluginDisplayName,
+      getSupportedAssets: async ({ direction, paymentTypes, regionCode }): Promise<FiatProviderAssetMap> => {
+        validateExactRegion(providerId, regionCode, SUPPORTED_REGIONS)
+        // Return nothing if paymentTypes are not supported by this provider
+        if (!paymentTypes.some(paymentType => allowedPaymentTypes[direction][paymentType] === true))
+          throw new FiatProviderError({ providerId, errorType: 'paymentUnsupported' })
+        const allowedCurrencyCodes = direction === 'buy' ? allowedBuyCurrencyCodes : allowedSellCurrencyCodes
+
+        if (Object.keys(allowedCurrencyCodes.crypto).length > 0) {
+          return allowedCurrencyCodes
+        }
+
+        const response = await fetch(`${urls.api[MODE]}/v1/ramp/blockchains`, {
+          method: 'GET',
+          headers: {
+            'Content-Type': 'application/json',
+            'X-Widget-Id': apiKey
+          }
+        })
+        if (!response.ok) {
+          const text = await response.text()
+          throw new Error(`Error fetching kado blockchains: ${text}`)
+        }
+        const result = await response.json()
+
+        const blockchains = asBlockchains(result)
+        if (!blockchains.success) {
+          throw new FiatProviderError({ providerId, errorType: 'paymentUnsupported' })
+        }
+
+        for (const blockchain of blockchains.data.blockchains) {
+          const { liveOnRamp } = blockchain
+          if (!liveOnRamp) continue
+          const pluginId = CHAIN_ID_TO_PLUGIN_MAP[blockchain.origin]
+          if (pluginId == null) continue
+          allowedCurrencyCodes.crypto[pluginId] = []
+          const tokens = allowedCurrencyCodes.crypto[pluginId]
+
+          for (const asset of blockchain.associatedAssets) {
+            const { isNative, address } = asset
+
+            if (asset.rampProducts == null || !asset.rampProducts.includes(direction)) continue
+            if (isNative) {
+              tokens.push({ tokenId: null, otherInfo: { symbol: asset.symbol } })
+              continue
+            }
+
+            if (address != null && address !== '0x0000000000000000000000000000000000000000') {
+              let tokenId: string
+              if (address.startsWith('0x')) {
+                // For EVM tokens only, lowercase and remove 0x
+                tokenId = address.toLowerCase().replace('0x', '')
+              } else {
+                tokenId = address
+              }
```

Comment:
Still not addressed. I don't mind this code if you provide some justification for it, but otherwise we should really try to avoid duplicating core logic in the GUI.